craft_application/services/service_factory.py

```diff
@@ -70,6 +71,10 @@ def __getattr__(self, service: str) -> services.BaseService:
         instantiated service as an instance attribute, allowing the same service
         instance to be reused for the entire run of the application.
         """
+        if self.project is None:  # pyright: ignore[reportUnnecessaryComparison]
+            raise RuntimeError(
+                "Provide a project to the ServiceFactory before starting services."
```

Comment:
This is intended for developers of craft applications, not end users?

Reply:
Yes. The only time a decently well-behaved application would have this pop up is if the developer overrides `Application._configure_services`, doesn't call `super()._configure_services()`, and doesn't set the factory's `project` in their `_configure_services`. 

Now that I've thought all of that through I'm going to replace the `RuntimeError` with some sort of `CraftError`, because there's no case where this isn't a bug in the application.

Reply:
OK, I've changed this over and temporarily used my comment above as the doc URL for the exception (until `craft-application` has actual docs about it).